packages/framework/src/Model/Product/Elasticsearch/MarkProductForExportSubscriber.php

```diff
@@ -0,0 +1,114 @@
+
+class MarkProductForExportSubscriber implements EventSubscriberInterface
+{
+    /**
+     * @var \Shopsys\FrameworkBundle\Model\Product\ProductFacade
+     */
+    protected $productFacade;
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Product\ProductFacade $productFacade
+     */
+    public function __construct(ProductFacade $productFacade)
+    {
+        $this->productFacade = $productFacade;
+    }
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Product\Availability\AvailabilityEvent $availabilityEvent
+     */
+    public function markAffectedByAvailability(AvailabilityEvent $availabilityEvent): void
+    {
+        $products = $this->productFacade->getProductsWithAvailability($availabilityEvent->getAvailability());
+        $this->productFacade->markProductsForExport($products);
+    }
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Product\Brand\BrandEvent $brandEvent
+     */
+    public function markAffectedByBrand(BrandEvent $brandEvent): void
+    {
+        $productIds = $this->productFacade->getProductsWithBrand($brandEvent->getBrand());
+        $this->productFacade->markProductsForExport($productIds);
+    }
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Product\Flag\FlagEvent $flagEvent
+     */
+    public function markAffectedByFlag(FlagEvent $flagEvent): void
+    {
+        $products = $this->productFacade->getProductsWithFlag($flagEvent->getFlag());
+        $this->productFacade->markProductsForExport($products);
+    }
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Product\Parameter\ParameterEvent $parameterEvent
+     */
+    public function markAffectedByParameter(ParameterEvent $parameterEvent): void
+    {
+        $products = $this->productFacade->getProductsWithParameter($parameterEvent->getParameter());
+        $this->productFacade->markProductsForExport($products);
+    }
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Product\Unit\UnitEvent $unitEvent
+     */
+    public function markAffectedByUnit(UnitEvent $unitEvent): void
+    {
+        $products = $this->productFacade->getProductsWithUnit($unitEvent->getUnit());
+        $this->productFacade->markProductsForExport($products);
+    }
+
+    /**
+     * @param \Symfony\Component\EventDispatcher\Event $event
+     */
+    public function markAll(Event $event): void
+    {
+        $this->productFacade->markAllProductsForExport();
+    }
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Component\Elasticsearch\IndexExportedEvent $indexExportedEvent
+     */
+    public function markAllAsExported(IndexExportedEvent $indexExportedEvent): void
+    {
+        if ($indexExportedEvent->getIndex() instanceof ProductIndex) {
+            $this->productFacade->markAllProductsAsExported();
+        }
+    }
+
+    /**
+     * @return array
+     */
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            ParameterEvent::DELETE => 'markAffectedByParameter',
```

Comment:
Or maybe just replace calling `ProductExportScheduler` by `$product->markForExport();`?

Reply:
It's a nice idea and surely it can help have the things unified. 
But I see one drawback here. Direct changes in a product will be propagated into the elastic later than it's now. 
Currently, the direct change of product is exported into elastic at the end of the request. But with this change, it would mean it can be in elastic after some time.

I prefer to not do this in sake of better UX. Do you agree?

Reply:
It is. The "why" is:
If I update a product directly, I can affect only one product (with its variants at the most). 
But when I, for example, change the name of availability, I can affect large number of products and this can be time-consuming to be done in one request.